src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java

```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
                 .orElseThrow(() -> new ResourceNotFoundException("Warehouse with id=" + warehouseId + " doesn't exist"));
     }
 
-    // TODO: 11/4/20
+    @Transactional
     @Override
-    public WarehouseDto save(WarehouseDto dto) {
-        return null;
+    public WarehouseDto save(Long customerId, WarehouseDto warehouseDto) {
+        if (warehouseDto.getIdentifier() == null) {
+            throw new ResponseStatusException(
+                    HttpStatus.NOT_ACCEPTABLE, "\"Identifier\" can't be null");
+        }
+
+        Warehouse warehouse = warehouseMapper.map(warehouseDto);
+        Address savedAddress = addressCreate(warehouseDto.getAddressDto());
+        warehouse.setAddress(savedAddress);
+        Optional<Customer> customer = customerRepository.findById(customerId);
+        customer.ifPresent(warehouse::setCustomer);
+        final Warehouse savedWarehouse = warehouseRepository.save(warehouse);
+
+        for (User user : warehouse.getUsers()) {
+            user.setWarehouse(warehouse);
+            userRepository.save(user);
+        }
+
+        return warehouseMapper.map(savedWarehouse);
     }
 
-    // TODO: 11/4/20
+    @Transactional
     @Override
-    public void delete(Long id) {
+    public WarehouseDto update(WarehouseDto warehouseDto) {
+        Warehouse entity = warehouseRepository.findById(warehouseDto.getId()).orElseThrow();
+        if (!warehouseDto.getIdentifier().equals(entity.getIdentifier())) {
+            throw new ResponseStatusException(
+                    HttpStatus.NOT_ACCEPTABLE, "You can't change \"identifier\" field");
+        }
```

Comment:
why can't we change identifier of warehouse? we only know that it should be unique.